### PR TITLE
[WIP] - Introduce socket factory

### PR DESCRIPTION
**Pull request by 13h3r**
Idea/API verification. Will update readme and scaladocs later

### CODE REVIEW

File: src/main/scala/ru/dgis/reactivezmq/ReactiveZMQ.scala [hunk 1]
```diff
@@ -49,8 +49,16 @@ object ZMQSource {
     * @param addresses a list of ZMQ endpoints to connect to
     * @return a Source of bytes
     */
-  def apply(socketFactory: => ZMQ.Socket, addresses: List[String]): Source[ByteString, Control] =
-    create(ZMQSocket(socketFactory), addresses)
+  def apply(socketFactory: () => ZMQ.Socket, addresses: List[String]): Source[ByteString, Control] =
+    create(() => ZMQSocket(socketFactory.apply()), addresses)
```
Comment (savulchik) [inline src/main/scala/ru/dgis/reactivezmq/ReactiveZMQ.scala]: socketFactory.andThen(ZMQSocket)?

File: src/main/scala/ru/dgis/reactivezmq/ReactiveZMQ.scala [hunk 2]
```diff
@@ -49,8 +49,16 @@ object ZMQSource {
...
+
+  def apply(context: ZMQ.Context, mode: Int, timeout: Int, addresses: List[String]): Source[ByteString, Control] = {
```
Comment (savulchik) [inline src/main/scala/ru/dgis/reactivezmq/ReactiveZMQ.scala]: we should add scaladoc for this constructor method

File: src/test/scala/ru/dgis/reactivezmq/Examples.scala [hunk 3]
```diff
@@ -14,7 +14,7 @@ object Examples extends App {
   val socket = context.socket(ZMQ.PULL)
   socket.setReceiveTimeOut(1000)
 
-  val source = ZMQSource(socket, List("tcp://127.0.0.1:12345"))
+  val source = ZMQSource(() => socket, List("tcp://127.0.0.1:12345"))
```
Comment (savulchik) [inline src/test/scala/ru/dgis/reactivezmq/Examples.scala]: i believe we should rewrite this example in order to show that you musn't reuse the same socket instance among materialized graphs

File: src/main/scala/ru/dgis/reactivezmq/ReactiveZMQ.scala (hunk 1 above)
Comment (13h3r) [inline src/main/scala/ru/dgis/reactivezmq/ReactiveZMQ.scala]: There is no `andThen` in `Function0`. Yes, I was shocked too.

**Comment by 13h3r** [comment]
Review, please

File: README.md [hunk 4]
```diff
@@ -27,19 +27,12 @@ resolvers += Resolver.jcenterRepo
...
 val context = ZMQ.context(1)
-val socket = context.socket(ZMQ.PULL)
-socket.setReceiveTimeOut(1000) // this should be >= 0
-```
-
-Create `Source` from `Socket`:
-
-```scala
-val source = ZMQSource(socket, List("tcp://127.0.0.1:12345"))
+val source = ZMQSource(context, ZMQ.PULL, 1 second, List("tcp://127.0.0.1:12345"))
```
Comment (savulchik) [inline README.md]: Let's use named parameters in this example:)

File: src/main/scala/ru/dgis/reactivezmq/ReactiveZMQ.scala [hunk 5]
```diff
@@ -50,8 +51,34 @@ object ZMQSource {
...
-  def apply(socketFactory: => ZMQ.Socket, addresses: List[String]): Source[ByteString, Control] =
-    create(ZMQSocket(socketFactory), addresses)
+  def apply(socketFactory: () => ZMQ.Socket, addresses: List[String]): Source[ByteString, Control] =
+    create(() => ZMQSocket(socketFactory.apply()), addresses)
+
+  /**
+    * Creates a ZMQ socket and wraps it with a Source
+    *
+    * The sockets from the factory:
+    *   -
```
Comment (savulchik) [inline src/main/scala/ru/dgis/reactivezmq/ReactiveZMQ.scala]: empty list item

File: src/main/scala/ru/dgis/reactivezmq/ReactiveZMQ.scala [hunk 6]
```diff
@@ -50,8 +51,34 @@ object ZMQSource {
...
+    *   - must have a non-negative receive timeout set
+    *
+    * The Source:
+    *   - emits when there is demand and the data available in the socket
+    *   - completes when graceful stop is initiated and the remaining data is delivered from the socket
+    *   - stops the delivery if downstream cancels the stream possibly loosing some data still remaining in the socket
+    *
+    * @param context context to create socket with
+    * @param mode socket type to be created. Must be ZMQ.PULL or ZMQ.SUB
+    * @param timeout context to create socket with
```
Comment (savulchik) [inline src/main/scala/ru/dgis/reactivezmq/ReactiveZMQ.scala]: wrong description

File: README.md (hunk 4 above)
Comment (13h3r) [inline README.md]: ok

File: src/main/scala/ru/dgis/reactivezmq/ReactiveZMQ.scala (hunk 5 above)
Comment (13h3r) [inline src/main/scala/ru/dgis/reactivezmq/ReactiveZMQ.scala]: ok

File: src/main/scala/ru/dgis/reactivezmq/ReactiveZMQ.scala (hunk 6 above)
Comment (13h3r) [inline src/main/scala/ru/dgis/reactivezmq/ReactiveZMQ.scala]: ok

**Pull request merged**